source/bitslab/src/index.rs

```diff
@@ -0,0 +1,84 @@
+use portable_atomic::{AtomicU16, Ordering::*};
+
+/// An allocator for up to 16 unique indices.
+pub struct IndexAlloc16(AtomicU16);
+
+impl IndexAlloc16 {
+    /// Returns a new allocator for up to 16 unique indices.
+    #[must_use]
+    pub const fn new() -> Self {
+        Self(AtomicU16::new(0))
+    }
+
+    /// Allocate an index from the pool.
+    ///
+    /// If this method returns [`Some`], the returned [`u8`] index will not be
+    /// returned again until after it has been [`free`](Self::free)d.
+    #[must_use]
+    pub fn allocate(&self) -> Option<u8> {
+        let mut bitmap = self.0.load(Acquire);
+        loop {
+            let idx = find_zero(bitmap)?;
+            let new_bitmap = bitmap | (1 << idx);
+            match self
+                .0
+                .compare_exchange_weak(bitmap, new_bitmap, AcqRel, Acquire)
+            {
+                Ok(_) => return Some(idx),
+                Err(actual) => bitmap = actual,
+            }
+        }
+    }
+
+    /// Release an index back to the pool.
+    ///
+    /// The freed index may now be returned by a subsequent call to [`allocate`](Self::allocate).
+    #[inline]
+    pub fn free(&self, index: u8) {
+        self.0.fetch_and(!(1 << index), Release);
+    }
+
+    /// Returns `true` if all indices in the allocator have been allocated.
```

Comment:
are the docs for full/empty backwards?

Reply:
no wait, this is just a reverse perspective thing, "full => no allocation"/"empty => all allocated". *might* be worth using something other than "full"/"empty", like `all_available`/`none_available`, but maybe I'm just silly.

Reply:
yeah, i went back and forth on what full and empty meant in this context. i think maybe i should change them to something more like what you suggested --- good call!